lgc/builder/BuilderImpl.cpp

```diff
@@ -329,6 +330,113 @@ BranchInst *BuilderImpl::createIf(Value *condition, bool wantElse, const Twine &
+    Instruction *operator*() {
+      unsigned index = *it;
+      assert(index < indexToInstr.size() && "Index out of range.");
+      return indexToInstr[index];
+    }
+
+    bool operator!=(const iterator &otherIt) {
+      assert(&otherIt.indexToInstr == &indexToInstr && "Iterators of different objects.");
+      return otherIt.it != it;
+    }
+  };
+
+  iterator begin(const IndexToInstructionVec &indexToInstr) const {
+    return iterator(bits.set_bits_begin(), indexToInstr);
+  }
+
+  iterator end(const IndexToInstructionVec &indexToInstr) const { return iterator(bits.set_bits_end(), indexToInstr); }
+
+  void insert(Instruction *instr, const InstrToIndexMap &instrToIndex) {
+    auto it = instrToIndex.find(instr);
+    assert(it != instrToIndex.end() && "Expected to find instr in instrToIndex.");
+    unsigned index = it->second;
+    bits.set(index);
+  }
+
+  unsigned size() const { return bits.size(); }
+
+  bool empty() const { return bits.empty(); }
+};
+
+class TraceNonUniformIndex {
+  // Maps the instruction to its index in the bit vector.
+  TinyInstructionSet::InstrToIndexMap instrToIndex;
+  // The instructions used as keys in instrToIndex in program order. It is used to map an index to an instruction.
+  TinyInstructionSet::IndexToInstructionVec indexToInstr;
+  // Maps an instruction to its dependencies.
+  DenseMap<Value *, TinyInstructionSet> instrDeps;
+  bool scalarizeDescriptorLoads;
+
+  void insertNewValueInInsnDeps(Value *, Instruction *);
+
+public:
+  TraceNonUniformIndex(Instruction *nonUniformInst, bool scalarizeDescriptorLoads = false)
+      : scalarizeDescriptorLoads(scalarizeDescriptorLoads) {
+    // Initialization of instrToIndex and indexToInstr.
+    if (scalarizeDescriptorLoads) {
+      unsigned cnt = 0;
+      for (Instruction *I = nonUniformInst->getPrevNode(); I != nullptr && cnt < 64; I = I->getPrevNode(), ++cnt) {
+        indexToInstr.push_back(I);
+        instrToIndex[I] = cnt;
+      }
+    }
+  }
+
+  Value *run(Value *);
+
+  const DenseMap<Value *, TinyInstructionSet> &getInstrDeps() const { return instrDeps; }
+
+  const TinyInstructionSet::IndexToInstructionVec &getIndexToInstr() const { return indexToInstr; }
+
+  bool foundDependencies() const { return scalarizeDescriptorLoads; }
+};
+
+void TraceNonUniformIndex::insertNewValueInInsnDeps(Value *newValue, Instruction *currentVisitedInsn) {
+  if (!instrToIndex.contains(currentVisitedInsn)) {
+    // The instruction is either outside of 64 limit or in a different basic block. So, we bail-out scalarization.
+    scalarizeDescriptorLoads = false;
+    return;
+  }
+  assert(instrDeps.contains(currentVisitedInsn) && "The current visited instruction should have been in the map.");
+  auto it1 = instrDeps.try_emplace(newValue).first;
+  auto &setOfInsns = it1->second;
+  auto it2 = instrDeps.find(currentVisitedInsn);
+  const auto &set = it2->second;
+  for (auto it3 = set.begin(indexToInstr), ite = set.end(indexToInstr); it3 != ite; ++it3) {
```

Comment:
Can you use a range based for loop instead?

Reply:
I did not do it because begin() and end() of TinyInstructionSet have an argument. Please let me know if there is a way to do it.

Reply:
I see, you originally stored indexToInstr as a member reference, right? With that approach, it would have worked

Reply:
Exactly :)

Reply:
How about not storing the reference to indexToInstr and instead add a separate member function that returns an `iterator_range` on which you can iterate?

````
iterator_range<IteratorT> getIteratorRange(IndexToInstrMap &indexToInstr) {
return iterator_range(set.begin(indexToInstr), set.end(indexToInstr));
}
```

That should allow you to make use of the range based for loop and not increase the storage size of the TinyInstructionSet?